lib/pure/algorithm.nim

```diff
@@ -559,6 +561,85 @@ proc isSorted*[T](a: openArray[T], order = SortOrder.Ascending): bool =
     assert isSorted(e) == false
   isSorted(a, system.cmp[T], order)
 
+proc merge*[T](
+  result: var seq[T],
+  x, y: openArray[T], cmp: proc(x, y: T): int {.closure.}
+) {.since: (1, 5, 1).} =
+  ## Merges two sorted `openArray`. `x` and `y` are assumed to be sorted.
+  ## If you do not wish to provide your own `cmp`,
+  ## you may use `system.cmp` or instead call the overloaded
+  ## version of `merge`, which uses `system.cmp`.
+  ##
+  ## **See also:**
+  ## * `merge proc<#merge,seq[T],openArray[T],openArray[T]>`_
+  runnableExamples:
+    let x = @[1, 3, 6]
+    let y = @[2, 3, 4]
+
+    var merged = @[7] # just to illustrate that its content will be discarded
```

Comment:
If the content of the `result` parameter will be discarded, then what's the point of having it? Why not make the signature `proc merge[T](x, y: openArray[T], cmp: proc(x, y: T): int {.closure.}): seq[T]`?

Reply:
> what's the point of having i

buffer preallocation.

But the real question is whether to append or clear + append. There are arguments on both sides; one argument for append is it's strictly more flexible; user can choose to setLen(0) before calling it.

Reply:
> buffer preallocation.

Hmm, I'd assume that in the majority of cases, it will be used like this
```nim
var s: seq[T]
merge(s, a, b)
```
and I'm not a huge fan of making the interface more awkward, just to optimize for an uncommon use case...

What about adding an overload that returns a new `seq`?

Reply:
this is what sugar.dup is for: avoiding duplicating each API with inplace vs outplace versions. All you need is inplace version. With dup you can use:
```nim
import std/[algorithm, sugar]
assert seq[int].default.dup(merge([1,3], [2,4])) == @[1,2,3,4] # with dup
assert merge(_,[1,3], [2,4])) == @[1,2,3,4] # pending https://github.com/nim-lang/RFCs/issues/292
```

Reply:
fixed